MatrixSDK/MXSession.m

```diff
@@ -3038,4 +3054,68 @@ - (void)notifyListeners:(MXEvent*)event direction:(MXTimelineDirection)direction
     }
 }
 
+#pragma mark - Publicised groups
+
+- (void)markOutdatedPublicisedGroupsByUserData
+{
+    // Retrieve the current list of users for who a publicised groups list is available.
+    // A server request will be triggered only when the publicised groups will be requested again for these users.
+    userIdsWithOutdatedPublicisedGroups = [NSMutableArray arrayWithArray:[publicisedGroupsByUserId allKeys]];
+}
+
+- (NSArray<NSString *> *)publicisedGroupsForUser:(NSString*)userId
+{
+    NSArray *publicisedGroups;
+    if (userId)
+    {
+        publicisedGroups = publicisedGroupsByUserId[userId];
+        
+        BOOL shouldRefresh = NO;
+        
+        if (!publicisedGroups)
+        {
+            shouldRefresh = YES;
+            
+            // In order to prevent multiple request on the same user id, we put a temporary empty array when no value is available yet.
+            // This temporary array will be replaced by the value received from the server.
+            publicisedGroups = publicisedGroupsByUserId[userId] = @[];
+        }
+        else if (userIdsWithOutdatedPublicisedGroups.count)
+        {
+            NSUInteger index = [userIdsWithOutdatedPublicisedGroups indexOfObject:userId];
+            if (index != NSNotFound)
+            {
+                shouldRefresh = YES;
+                
+                // Remove this user id from the pending list
+                [userIdsWithOutdatedPublicisedGroups removeObjectAtIndex:index];
+            }
+        }
+        
+        if (shouldRefresh)
+        {
+            [self.matrixRestClient getPublicisedGroupsForUsers:@[userId] success:^(NSDictionary<NSString *,NSArray<NSString *> *> *updatedPublicisedGroupsByUserId) {
+                
+                // Check whether the publicised groups have been actually modified.
+                if ((publicisedGroupsByUserId[userId] || updatedPublicisedGroupsByUserId[userId]) && ![publicisedGroupsByUserId[userId] isEqualToArray:updatedPublicisedGroupsByUserId[userId]])
+                {
+                    // refresh the internal dict
+                    publicisedGroupsByUserId[userId] = updatedPublicisedGroupsByUserId[userId];
+                    
+                    // Notify the publicised groups for these users have been updated.
+                    [[NSNotificationCenter defaultCenter] postNotificationName:kMXSessionDidUpdatePublicisedGroupsForUsersNotification
+                                                                        object:self
+                                                                      userInfo:@{
+                                                                                 kMXSessionNotificationUserIdsArrayKey: @[userId]
+                                                                                 }];
+                }
+                
+                
+            } failure:nil];
```

Comment:
In case of failure, it may worth to put `userId` in `userIdsWithOutdatedPublicisedGroups`, no?

Reply:
yes